File: src/main/scala/org/xbery/artbeams/common/antispam/repository/AntispamQuizMapper.scala
package org.xbery.artbeams.common.antispam.repository

import org.xbery.artbeams.common.antispam.domain.{AntispamQuiz, AntispamQuizFilter}
import org.xbery.overview.filter.Condition
import org.xbery.overview.mapper._
import org.xbery.overview.repo.Conditions

import java.util

/**
  * Maps {@link AntispamQuiz} entity to set of attributes and vice versa.
  * @author Radek Beran
  */
class AntispamQuizMapper() extends DynamicEntityMapper[AntispamQuiz, AntispamQuizFilter] {

  private val cls = classOf[AntispamQuiz]

  override val getTableName: String = "antispam_quiz"

  val questionAttr = add(Attr.ofString(cls, "question").get(e => e.question).primary())
  val answerAttr = add(Attr.ofString(cls, "answer").get(e => e.answer))

  override def createEntity(attributeSource: AttributeSource, attributes: java.util.List[Attribute[AntispamQuiz, _]], aliasPrefix: String): AntispamQuiz = {
    AntispamQuiz(
      questionAttr.getValueFromSource(attributeSource, aliasPrefix),
      answerAttr.getValueFromSource(attributeSource, aliasPrefix)
    )
  }

  override def composeFilterConditions(filter: AntispamQuizFilter): util.List[Condition] = {
    val conditions = new util.ArrayList[Condition]
    filter.question.map(question => conditions.add(Conditions.eq(this.questionAttr, question)))
    conditions
  }
}

object AntispamQuizMapper {
  lazy val Instance = new AntispamQuizMapper()
}

File: src/main/kotlin/org/xbery/artbeams/common/antispam/repository/AntispamQuizMapper.kt
package org.xbery.artbeams.common.antispam.repository

import org.xbery.artbeams.common.antispam.domain.AntispamQuiz
import org.xbery.artbeams.common.antispam.domain.AntispamQuizFilter
import org.xbery.overview.filter.Condition
import org.xbery.overview.mapper.Attr
import org.xbery.overview.mapper.Attribute
import org.xbery.overview.mapper.AttributeSource
import org.xbery.overview.mapper.DynamicEntityMapper
import org.xbery.overview.repo.Conditions

/**
 * Maps {@link AntispamQuiz} entity to set of attributes and vice versa.
 * @author Radek Beran
 */
open class AntispamQuizMapper() : DynamicEntityMapper<AntispamQuiz, AntispamQuizFilter>() {
    private val cls: Class<AntispamQuiz> = AntispamQuiz::class.java
    override fun getTableName() = "antispam_quiz"
    val questionAttr: Attribute<AntispamQuiz, String> =
        add(Attr.ofString(cls, "question").get { e -> e.question }.primary())
    val answerAttr: Attribute<AntispamQuiz, String> =
        add(Attr.ofString(cls, "answer").get { e -> e.answer })

    override fun createEntity(
        attributeSource: AttributeSource,
        attributes: List<Attribute<AntispamQuiz, *>>,
        aliasPrefix: String?
    ): AntispamQuiz {
        return AntispamQuiz(
            questionAttr.getValueFromSource(
                attributeSource,
                aliasPrefix
            ),
            answerAttr.getValueFromSource(attributeSource, aliasPrefix ?: "")
        )
    }

    override fun composeFilterConditions(filter: AntispamQuizFilter): MutableList<Condition> {
        val conditions = mutableListOf<Condition>()
        filter.question?.let { question -> conditions.add(Conditions.eq(this.questionAttr, question)) }
        return conditions
    }

    companion object {
        val Instance: AntispamQuizMapper = AntispamQuizMapper()
    }
}
